Failure during creation of plan fragments: Cannot overwrite partitioning with SINGLE (currently set to Optional[HASH])

Saw an issue which is reproducible with the following sequence

```sql
CREATE TABLE f1 (k1 bigint, k2 bigint);
CREATE TABLE f2_p (k1 bigint, k2 bigint) WITH (partitioned_by = array['k2']);
CREATE TABLE f3 (k1 bigint, k2 bigint);
INSERT INTO f1 VALUES (5, 6);
INSERT INTO f2_p VALUES (5, 6);
INSERT INTO f3 VALUES (5, 6);


WITH t1 AS (
  SELECT 
    k1 
  FROM (
    SELECT 
      k1,
      ROW_NUMBER() OVER (PARTITION BY k1) AS some_rank
    FROM f1
  )
  WHERE 
    some_rank = 1
  ),
t2 AS (
  SELECT k1 FROM f2_p WHERE k2 - 5 = 8)
SELECT 
  t3.k1 FROM f3 AS t3
  INNER JOIN t1 on t3.k1 = t1.k1 
  INNER JOIN t2 ON t3.k1 = t2.k1; 
```


Query Plan

```
Output[k1]
│   Layout: [k1:bigint]
│   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
└─ RemoteExchange[GATHER]
   │   Layout: [k1:bigint]
   │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
   └─ InnerJoin[(""k1"" = ""k1_5"")][$hashvalue, $hashvalue_13]
      │   Layout: [k1:bigint]
      │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │   Distribution: PARTITIONED
      │   maySkipOutputDuplicates = false
      ├─ InnerJoin[(""k1"" = ""k1_0"")][$hashvalue, $hashvalue_11]
      │  │   Layout: [k1:bigint, $hashvalue:bigint]
      │  │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │  │   Distribution: PARTITIONED
      │  │   maySkipOutputDuplicates = false
      │  ├─ RemoteExchange[REPARTITION][$hashvalue]
      │  │  │   Layout: [k1:bigint, $hashvalue:bigint]
      │  │  │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │  │  └─ Project[]
      │  │     │   Layout: [k1:bigint, $hashvalue_10:bigint]
      │  │     │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │  │     │   $hashvalue_10 := combine_hash(bigint '0', COALESCE(""$operator$hash_code""(""k1""), 0))
      │  │     └─ Values
      │  │            Layout: [k1:bigint]
      │  │            Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │  └─ LocalExchange[HASH][$hashvalue_11] (""k1_0"")
      │     │   Layout: [k1_0:bigint, $hashvalue_11:bigint]
      │     │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │     └─ Project[]
      │        │   Layout: [k1_0:bigint, $hashvalue_12:bigint]
      │        │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      │        │   $hashvalue_12 := combine_hash(bigint '0', COALESCE(""$operator$hash_code""(""k1_0""), 0))
      │        └─ Values
      │               Layout: [k1_0:bigint]
      │               Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
      └─ LocalExchange[HASH][$hashvalue_13] (""k1_5"")
         │   Layout: [k1_5:bigint, $hashvalue_13:bigint]
         │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
         └─ RemoteExchange[REPARTITION][$hashvalue_14]
            │   Layout: [k1_5:bigint, $hashvalue_14:bigint]
            │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
            └─ Project[]
               │   Layout: [k1_5:bigint, $hashvalue_15:bigint]
               │   Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
               │   $hashvalue_15 := combine_hash(bigint '0', COALESCE(""$operator$hash_code""(""k1_5""), 0))
               └─ Values
                      Layout: [k1_5:bigint]
                      Estimates: {rows: 0 (0B), cpu: 0, memory: 0B, network: 0B}
```

Failure in PlanFragmenter

```
Caused by: java.lang.IllegalStateException: Cannot overwrite partitioning with SINGLE (currently set to Optional[HASH])
	at com.google.common.base.Preconditions.checkState(Preconditions.java:821)
	at io.trino.sql.planner.PlanFragmenter$FragmentProperties.setSingleNodeDistribution(PlanFragmenter.java:446)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitValues(PlanFragmenter.java:351)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitValues(PlanFragmenter.java:202)
	at io.trino.sql.planner.plan.ValuesNode.accept(ValuesNode.java:131)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:37)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:21)
	at io.trino.sql.planner.plan.PlanVisitor.visitProject(PlanVisitor.java:39)
	at io.trino.sql.planner.plan.ProjectNode.accept(ProjectNode.java:82)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitExchange(PlanFragmenter.java:359)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitExchange(PlanFragmenter.java:202)
	at io.trino.sql.planner.plan.ExchangeNode.accept(ExchangeNode.java:243)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:37)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:21)
	at io.trino.sql.planner.plan.PlanVisitor.visitJoin(PlanVisitor.java:99)
	at io.trino.sql.planner.plan.JoinNode.accept(JoinNode.java:316)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitExchange(PlanFragmenter.java:359)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitExchange(PlanFragmenter.java:202)
	at io.trino.sql.planner.plan.ExchangeNode.accept(ExchangeNode.java:243)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:37)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:21)
	at io.trino.sql.planner.plan.PlanVisitor.visitJoin(PlanVisitor.java:99)
	at io.trino.sql.planner.plan.JoinNode.accept(JoinNode.java:316)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitOutput(PlanFragmenter.java:266)
	at io.trino.sql.planner.PlanFragmenter$Fragmenter.visitOutput(PlanFragmenter.java:202)
	at io.trino.sql.planner.plan.OutputNode.accept(OutputNode.java:83)
	at io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:31)
	at io.trino.sql.planner.PlanFragmenter.createSubPlans(PlanFragmenter.java:125)
	at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:473)
	at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:450)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:391)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:249)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_testversion____20220927_032348_71.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

The issue is reproducible on current master, as well as as far back as 352

--- a/core/trino-main/src/main/java/io/trino/sql/planner/PlanFragmenter.java
+++ b/core/trino-main/src/main/java/io/trino/sql/planner/PlanFragmenter.java
@@ -347,7 +347,11 @@ public class PlanFragmenter
         @Override
         public PlanNode visitValues(ValuesNode node, RewriteContext<FragmentProperties> context)
         {
-            context.get().setSingleNodeDistribution();
+            // An empty values node is compatible with any distribution, so
+            // don't attempt to overwrite one's already been chosen
+            if (node.getRowCount() != 0 || !context.get().hasDistribution()) {
+                context.get().setSingleNodeDistribution();
+            }
             return context.defaultRewrite(node, context.get());
         }
 
@@ -433,6 +437,11 @@ public class PlanFragmenter
         public List<SubPlan> getChildren()
         {
             return children;
+        }
+
+        public boolean hasDistribution()
+        {
+            return partitioningHandle.isPresent();
         }
 
         public FragmentProperties setSingleNodeDistribution()

--- a/plugin/trino-hive/pom.xml
+++ b/plugin/trino-hive/pom.xml
@@ -431,6 +431,24 @@
         </dependency>
 
         <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
             <groupId>org.openjdk.jmh</groupId>
             <artifactId>jmh-core</artifactId>
             <scope>test</scope>
@@ -454,6 +472,28 @@
             <scope>test</scope>
         </dependency>
     </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <dependencies>
+                    <!-- allow both JUnit and TestNG -->
+                    <dependency>
+                        <groupId>org.apache.maven.surefire</groupId>
+                        <artifactId>surefire-junit-platform</artifactId>
+                        <version>${dep.plugin.surefire.version}</version>
+                    </dependency>
+                    <dependency>
+                        <groupId>org.apache.maven.surefire</groupId>
+                        <artifactId>surefire-testng</artifactId>
+                        <version>${dep.plugin.surefire.version}</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
+        </plugins>
+    </build>
 
     <profiles>
         <profile>

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestIssue14317.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestIssue14317.java
@@ -1,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive;
+
+import io.trino.sql.query.QueryAssertions;
+import io.trino.testing.DistributedQueryRunner;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Test for https://github.com/trinodb/trino/issues/14317
+ * <p>
+ * The only known reproduction involves a connector that exposes partitioned tables
+ * and supports predicate pushdown.
+ */
+public class TestIssue14317
+{
+    @Test
+    public void test()
+            throws Exception
+    {
+        HiveQueryRunner.Builder<?> builder = HiveQueryRunner.builder()
+                .setCreateTpchSchemas(false);
+
+        try (DistributedQueryRunner queryRunner = builder.build();
+                QueryAssertions assertions = new QueryAssertions(queryRunner);) {
+            queryRunner.execute("CREATE SCHEMA s");
+
+            queryRunner.execute("CREATE TABLE s.t (a bigint, b bigint)");
+            queryRunner.execute("CREATE TABLE s.u (c bigint, d bigint) WITH (partitioned_by = array['d'])");
+            queryRunner.execute("CREATE TABLE s.v (e bigint, f bigint)");
+
+            queryRunner.execute("INSERT INTO s.t VALUES (5, 6)");
+            queryRunner.execute("INSERT INTO s.u VALUES (5, 6)");
+            queryRunner.execute("INSERT INTO s.v VALUES (5, 6)");
+
+            assertThat(assertions.query("""
+                    WITH t1 AS (
+                      SELECT a
+                      FROM (
+                        SELECT a, ROW_NUMBER() OVER (PARTITION BY a) AS rn
+                        FROM s.t)
+                      WHERE rn = 1),
+                    t2 AS (SELECT c FROM s.u WHERE d - 5 = 8)
+                    SELECT v.e
+                    FROM s.v
+                      INNER JOIN t1 on v.e = t1.a
+                      INNER JOIN t2 ON v.e = t2.c
+                    """))
+                    .returnsEmptyResult();
+
+            queryRunner.execute("DROP TABLE s.t");
+            queryRunner.execute("DROP TABLE s.u");
+            queryRunner.execute("DROP TABLE s.v");
+            queryRunner.execute("DROP SCHEMA s");
+        }
+    }
+}

